src/srsnv/ugbio_srsnv/srsnv_training_utils.py

```diff
@@ -0,0 +1,1314 @@
+        all_train_cond = pd.Series(False, index=df.index)  # for keeping track of all "train" indices
+    for k in range(k_folds):
+        val_cond = df[kfold_col] == k
+        all_val_cond = np.logical_or(all_val_cond, val_cond)
+        if val_cond.any():
+            preds[val_cond] = models[k].predict_proba(df.loc[val_cond, columns_for_training], **kwargs)[:, 1]
+        if test_cond.any():
+            test_scores = prob_to_phred(
+                models[k].predict_proba(df.loc[test_cond, columns_for_training], **kwargs)[:, 1]
+            )
+            preds[test_cond] += test_scores / k_folds
+        if return_train:
+            train_cond = np.logical_and(~val_cond, ~test_cond)
+            all_train_cond = np.logical_or(all_train_cond, train_cond)
+            if train_cond.any():
+                train_scores = prob_to_phred(
+                    models[k].predict_proba(df.loc[train_cond, columns_for_training], **kwargs)[:, 1]
+                )
+                preds_train[train_cond] += train_scores / preds_train_norm
+
+    preds[test_cond] = phred_to_prob(preds[test_cond])
+    preds[~all_val_cond] = np.nan
+    if return_train:
+        preds_train = phred_to_prob(preds_train)
+        preds_train[~all_train_cond] = np.nan
+        preds = pd.concat((preds, preds_train), axis=1)
+
+    return preds
+
+
+def get_quality_interpolation_function(mrd_simulation_dataframe: str = None):
+    """
+    Create a function that interpolates between residual_snv_rate and ML_QUAL
+
+    Parameters
+    ----------
+    mrd_simulation_dataframe : str, None
+        Path to MRD simulation dataframe
+        If None then None is returned and nothing is done
+
+    Returns
+    -------
+    interp1d
+        Interpolation function
+    """
+    if not mrd_simulation_dataframe:
+        return None
+    # read data, filter for ML_QUAL filters only
+    df = pd.read_parquet(mrd_simulation_dataframe)
+    df.index = df.index.str.upper()
+    df = df[df.index.str.startswith(ML_QUAL)]
+    df.loc[:, ML_QUAL] = df.index.str.replace(ML_QUAL + "_", "").astype(float)
+    # Calculate Phred scores matching the residual SNV rate
+    df = df[df["residual_snv_rate"] > 0]  # Estimate using only residual SNV rates > 0 to avoid QUAL=Inf
+    phred_residual_snv_rate = -10 * np.log10(df["residual_snv_rate"])
+    # Create interpolation function
+    quality_interpolation_function = interp1d(
+        df[ML_QUAL] + 1e-10,  # add a small number so ML_QUAL=0 is outside the interpolation range
+        phred_residual_snv_rate,
+        kind="linear",
+        bounds_error=False,
+        fill_value=(
+            0,
+            phred_residual_snv_rate[-1],
+        ),  # below ML_QUAL=1E-10 assign 0, above max assign max
+    )
+    return quality_interpolation_function
+
+
+# pylint:disable=too-many-arguments
+# pylint:disable=too-many-branches
+# pylint:disable=too-many-statements
+def prepare_featuremap_for_model(
```

Comment:
wow this function is terrbile... It should ne splited to smaller functios, but this looks like a lot of work and not mandatory for now